fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/synchronizedsession/SynchronizedSession.java

```diff
@@ -0,0 +1,171 @@
+
+    public SynchronizedSessionRunner reuseSessionAndCreateRunner(@Nonnull UUID sessionId) {
+        this.sessionId = sessionId;
+        return new SynchronizedSessionRunner(runner, this);
+    }
+
+    // Return true if get lease otherwise false.
+    private CompletableFuture<Boolean> initializeSession(@Nonnull FDBRecordContext context) {
+        if (sessionId != null) {
+            throw new RecordCoreException("SynchronizedSession has been initialized");
+        }
+        sessionId = UUID.randomUUID();
+        Transaction tr = context.ensureActive();
+        return getLockSessionId(tr).thenCompose(lockSessionId -> {
+            if (lockSessionId == null) {
+                // If there was no lock, can get the lock.
+                return getSessionLock(tr);
+            } else if (lockSessionId == sessionId) {
+                // This should never happen.
+                throw new RecordCoreException("session id already exists");
+            } else {
+                // This is snapshot read so it will not affect all working transactions writing to it.
+                return getLockSessionTime(tr.snapshot()).thenCompose(sessionTime -> {
+                    long currentTime = System.currentTimeMillis();
+                    // TODO: Deal with clock skews.
+                    if (sessionTime + LEASE_PERIOD_MILL < currentTime) {
+                        // The old lease was outdated, can get the lock.
+                        return getSessionLock(tr);
+                    } else {
+                        return AsyncUtil.READY_FALSE;
+                    }
+                });
+            }
+        });
+    }
+
+    <T> Function<FDBRecordContext, T> workInSession(
+            @Nonnull Function<? super FDBRecordContext, ? extends T> work) {
+        return context -> context.asyncToSync(FDBStoreTimer.Waits.WAIT_RETRY_DELAY,
+                workInSessionAsync(ctx -> CompletableFuture.completedFuture(work.apply(ctx))).apply(context));
+    }
+
+    <T> Function<? super FDBRecordContext, CompletableFuture<? extends T>> workInSessionAsync(
+            @Nonnull Function<? super FDBRecordContext, CompletableFuture<? extends T>> work) {
+        return context -> getLockSessionId(context.ensureActive())
+                .thenCompose(lockSessionId -> {
+                    if (lockSessionId != sessionId) {
+                        throw new SynchronizedSessionExpiredException("failed to continue the session");
+                    }
+                    return AsyncUtil.DONE;
+                })
+                .thenCompose(vignore -> work.apply(context))
+                .thenApply(result -> {
+                    setLockSessionTime(context.ensureActive());
+                    return result;
+                });
+    }
+
+    private CompletionStage<Boolean> getSessionLock(Transaction tr) {
+        setLockSessionId(tr);
+        setLockSessionTime(tr);
+        return tr.commit().thenCompose(vignore -> AsyncUtil.READY_TRUE);
```

Comment:
Do we want to commit immediately upon assuming we have a lock?  I guess this is kind of infrequent, so maybe that is ok